compiler/AST/ForallStmt.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ForallStmt.h"
+#include "AstVisitor.h"
+#include "passes.h"
+
+/////////////////////////////////////////////////////////////////////////////
+//
+// ForallStmt represents a forall loop statement
+//
+/////////////////////////////////////////////////////////////////////////////
+
+ForallStmt::ForallStmt(bool zippered, BlockStmt* body, ForallIntents* with):
+  Stmt(E_ForallStmt),
+  fZippered(zippered),
+  fLoopBody(body),
+  fWith(with)
+{
+  fIterVars.parent = this;
+  fIterExprs.parent = this;
+  fIntentVars.parent = this;
+  gForallStmts.add(this);
+}
+
+ForallStmt::~ForallStmt() {
+  delete fWith;
+}
+
+ForallStmt* ForallStmt::copyInner(SymbolMap* map) {
+  ForallStmt* _this  = new ForallStmt(fZippered,
+                                      COPY_INT(fLoopBody),
+                                      COPY_INT(fWith));
+  for_alist(expr, fIterVars)
+    _this->fIterVars.insertAtTail(COPY_INT(expr));
+  for_alist(expr, fIterExprs)
+    _this->fIterExprs.insertAtTail(COPY_INT(expr));
+  for_alist(expr, fIntentVars)
+    _this->fIntentVars.insertAtTail(COPY_INT(expr));
+
+  return _this;
+}
+
+void ForallStmt::replaceChild(Expr* oldAst, Expr* newAst) {
+  if (oldAst == fLoopBody) {
+    if (!newAst)
+      fLoopBody = NULL;
+    else if (BlockStmt* newBlock = toBlockStmt(newAst))
+      fLoopBody = newBlock;
+    else
+      // It is caller responsibility to make newAst fit.
+      INT_ASSERT(false);
+
+  } else if (fWith->replaceChildFI(oldAst, newAst))
+    ; //good
```

Comment:
a nit, I think I'd prefer to use curly braces in this situation instead of a bare ;